### PR TITLE
Pytest

### PR DESCRIPTION
**Pull request by RomainBrault**
#30

### CODE REVIEW

File: ci_scripts/travis/install.sh [hunk 1]
```diff
@@ -24,7 +24,7 @@ popd
 
 # Configure the conda environment and put it in the path using the
 # provided versions
-conda create -n testenv --yes python=$PYTHON_VERSION pip nose \
+conda create -n testenv --yes python=$PYTHON_VERSION pip pytest pytest-cov \
```
Comment (rth) [inline ci_scripts/travis/install.sh]: It might still be worth keeping nose in the dependencies (in addition to pytest), as e.g. older versions of `numpy.testing` will use it, if I remember correctly.

**Review by rth — COMMENTED** [review]
Thanks for this PR! 

* Please also replace as needed [other mentions of `nosetests`](https://github.com/scikit-learn-contrib/project-template/search?utf8=%E2%9C%93&q=nosetests&type=) in particular in the readme and in `setup.cfg`.

 * You can also replace `nose` setup by `pytest` in `appveyor.yml`

**Comment by RomainBrault** [comment]
I kept nose in install.sh and appveyor.yml  and .gitignore; and added pytest to install.sh and appveyor.yml. No other mentions of nose should be present.

**Comment by RomainBrault** [comment]
Also, I'm wondering why requirements.txt includes nose? I think neither pytest or nose should be in.

File: .gitignore [hunk 2]
```diff
@@ -47,6 +47,7 @@ htmlcov/
 .coverage
 .coverage.*
 .cache
+.test.sh.swp
```
Comment (rth) [inline .gitignore]: That's not necessary, it's specific to your editor (vim I imagine) and should be set in your [global .gitignore](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c)

File: ci_scripts/travis/test.sh [hunk 3]
```diff
@@ -7,7 +7,8 @@ mkdir -p $TEST_DIR
 cd $TEST_DIR
 
 if [[ "$COVERAGE" == "true" ]]; then
-    nosetests -s --with-coverage --cover-package=$MODULE $MODULE
+    py.test -v -x --doctest-modules --ignore=setup.py --pyargs $MODULE \
+        --cov=$MODULE
 else
-    nosetests -s $MODULE
+    py.test -v -x --doctest-modules --ignore=setup.py --pyargs $MODULE
```
Comment (rth) [inline ci_scripts/travis/test.sh]: `--ignore` doesn't work with `--pyargs` unfortunately https://github.com/pytest-dev/pytest/issues/3287

**Review by rth — COMMENTED** [review]
Thanks for the update; a few more comments are below.

>> It might still be worth keeping nose in the dependencies (in addition to pytest), as e.g. older versions of numpy.testing will use it, if I remember correctly.
>
> I kept nose in install.sh and appveyor.yml and .gitignore; and added pytest to install.sh and appveyor.yml. No other mentions of nose should be present.

I'm not sure what should be done there; maybe @lesteve would know...

File: ci_scripts/travis/test.sh [hunk 4]
```diff
@@ -7,8 +7,7 @@ mkdir -p $TEST_DIR
 cd $TEST_DIR
 
 if [[ "$COVERAGE" == "true" ]]; then
-    py.test -v -x --doctest-modules --ignore=setup.py --pyargs $MODULE \
-        --cov=$MODULE
+    py.test -v -x --doctest-modules --ignore=setup.py --cov=$MODULE
 else
-    py.test -v -x --doctest-modules --ignore=setup.py --pyargs $MODULE
+    py.test -v -x --doctest-modules --ignore=setup.py
```
Comment (RomainBrault) [inline ci_scripts/travis/test.sh]: I was a bit confused. Actually pyargs is mandatory to get the path right easily and ignore is superfluous

**Comment by lesteve** [comment]
> It might still be worth keeping nose in the dependencies (in addition to pytest), as e.g. older versions of numpy.testing will use it, if I remember correctly.

Very unfortunately scikit-learn 0.19.1 has a hard dependency on nose (try importing sklearn.utils.testing without nose installed) so it may be the case that we still need nose (as a dependency not to run the tests) in the CIs for scikit-learn-contrib.

numpy.testing is irrelevant, basically nose was still used in a weird edge case (maybe numpy had something like numpy.testing.assert_equal that was using nose). The thing is that there is no actual need to use numpy.testing.assert_equal (or whatever the actual function was) and you can emulate it like we do in scikit-learn with unittest.

**Comment by NicolasHug** [comment]
Is there anything keeping from merging this?
It could make the badges go green again :)

**Comment by rth** [comment]
> Is there anything keeping from merging this?

Finding someone who has merge rights probably..

cc @glemaitre

**Comment by NicolasHug** [comment]
@amueller ?

**Comment by glemaitre** [comment]
Merging this one. I think that we should have more update regarding this template.

**Comment by rth** [comment]
Thanks @glemaitre ! A few thoughts on what could be simplified in https://github.com/scikit-learn-contrib/project-template/issues/40 ..

**Comment by glemaitre** [comment]
I am working on it. I will make a todo list there.
